presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpGenericPattern.java

```diff
@@ -0,0 +1,435 @@
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+
+import java.nio.charset.StandardCharsets;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_SESSION_PROPERTY;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static java.lang.String.format;
+
+public final class RegexpGenericPattern
+{
+    enum RegexLibrary {
+        RE2J,
+        JONI
+    }
+
+    public Pattern re2jPattern;
+    public Regex joniRegex;
+    public final RegexLibrary selectedRegexLibrary;
+
+    public RegexpGenericPattern(String regexLibraryName)
+    {
+        if (regexLibraryName.toUpperCase().equals("RE2J")) {
+            selectedRegexLibrary = RegexLibrary.RE2J;
+        }
+        else if (regexLibraryName.toUpperCase().equals("JONI")) {
+            selectedRegexLibrary = RegexLibrary.JONI;
+        }
+        else {
+          throw new PrestoException(INVALID_SESSION_PROPERTY, "The setting for regex_library is not valid. It should be either RE2J or JONI.");
+        }
+    }
+
+    public void createPatternObject(Slice pattern) throws Exception
+    {
+        if (selectedRegexLibrary == RegexLibrary.RE2J) {
+            re2jPattern =  Pattern.compile(pattern.toStringUtf8());
+        }
+        else if (selectedRegexLibrary == RegexLibrary.JONI) {
+            joniRegex = new Regex(pattern.getBytes(), 0, pattern.length(), Option.DEFAULT, NonStrictUTF8Encoding.INSTANCE, Syntax.Java);
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        if (selectedRegexLibrary == RegexLibrary.RE2J) {
+            com.google.re2j.Matcher m = re2jPattern.matcher(source.toStringUtf8()); //sourceString);
+            return m.find();
+        }
+        else if (selectedRegexLibrary == RegexLibrary.JONI) {
+            io.airlift.joni.Matcher m = joniRegex.matcher(source.getBytes());
+            int offset = m.search(0, source.length(), Option.DEFAULT);
+            return offset != -1;
+        }
+        return false;
+    }
+
+    private void re2jCheckReplacementString(String replacementString, int groupCount)
+    {
+        //checking some special cases in the replacement string to make it compatible to joni processing
+        //Currently only handles exception cases similar to joni processing
+        int idx = 0;
+        int replacementStringLength = replacementString.length();
+        while (idx < replacementStringLength) {
+            char nextChar = replacementString.charAt(idx);
+            if (nextChar == '$') {
+                idx++;
+                if (idx == replacementStringLength) {
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacementString);
+                }
+                nextChar = replacementString.charAt(idx);
+                if (nextChar != '{') {
+                    int backref = nextChar - '0';
```

Comment:
backreferences are not supported by RE2J at all, right? So any `$n` should fail probably